engine/evarutil.ml

```diff
@@ -684,15 +685,39 @@ let undefined_evars_of_named_context evd nc =
     nc
     ~init:Evar.Set.empty
 
-let undefined_evars_of_evar_info evd evi =
+(* compute undefined evars, cache results *) 
+let cache_undefined_evars_of_evar_info ev evd evi =
+  let deps =
   Evar.Set.union (undefined_evars_of_term evd (EConstr.of_constr evi.evar_concl))
     (Evar.Set.union
        (match evi.evar_body with
 	 | Evar_empty -> Evar.Set.empty
-	 | Evar_defined b -> undefined_evars_of_term evd (EConstr.of_constr b))
+	 | Evar_defined b ->
+	    (* we assume ev is undefined, so body should not be defined *)
+	    assert false)
        (undefined_evars_of_named_context evd
-	  (named_context_of_val evi.evar_hyps)))
-
+          (named_context_of_val evi.evar_hyps)))
+  in
+  let evi' = { evi with evar_dependency_cache = Some deps } in
+  let evd' = Evd.update_undefined evd ev evi' in
+  (evi',evd')
+
+(* get undefined evars, from cache if possible *)
+let undefined_evars_of_evar_info ev evd evi =
+  let (evi_with_cache,evd_with_cache) =
+    match evi.evar_dependency_cache with
+    | None -> (* initially populate dependency cache *)
+       cache_undefined_evars_of_evar_info ev evd evi
+  | Some cache -> (* update dependency cache *)
+     (* some evars may have become defined, so repopulate cache *)
+     let cache_invalid = Evar.Set.exists (fun evar -> Evd.is_defined evd evar) cache in
```

Comment:
We could be a lot more clever when repopulating the cache and just remove the defined evars evs and extend it with the undefined evars of evs.

Reply:
That's what I thought at first, but this is actually unsound. We discussed that with @psteckler somewhere, probably on gitter.

Reply:
What can co wrong if we assume monotonous extension of the evd?